include/ensmallen_bits/adam/padam_update.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file padam_update.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of Partially adaptive momentum estimation method (Padam).
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_ADAM_PADAM_UPDATE_HPP
+#define ENSMALLEN_ADAM_PADAM_UPDATE_HPP
+
+namespace ens {
+
+/**
+ * Partially adaptive momentum estimation method (Padam),
+ * adopts historical gradient information to automatically adjust the
+ * learning rate.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{
+ *   title   = {Closing the Generalization Gap of Adaptive Gradient Methods in
+ *              Training Deep Neural Networks},
+ *   author  = {{Chen}, J. and {Gu}, Q.},
+ *   journal = {ArXiv e-prints},
+ *   url     = {https://arxiv.org/abs/1806.06763}
+ *   year    = {2018}
+ * }
+ * @endcode
+ */
+class PadamUpdate
+{
+ public:
+  /**
+   * Construct the Padam update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   * @param partial Partially adaptive parameter.
+   */
+  PadamUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
+              const double beta2 = 0.999,
+              const double partial = 0.25) :
+      epsilon(epsilon),
+      beta1(beta1),
+      beta2(beta2),
+      partial(partial),
+      iteration(0)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    v = arma::zeros<arma::mat>(rows, cols);
+    vImproved = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for Padam.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
```

Comment:
Do we also need to update `beta1` as `beta1 *= lambda` like the theory for Theorem 4.2 requires for convergence?  I guess that Algorithm 1 does not include it, so perhaps it is not necessary.  It also seems like they did not decay `beta1` in the experiments.

Reply:
Sounds good to me.  It can always be changed later if someone wants.